src/plugins/expressions/common/expression_functions/specs/map_column.ts

```diff
@@ -80,50 +80,50 @@ export const mapColumn: ExpressionFunctionDefinition<
       default: null,
     },
   },
-  fn: (input, args) => {
-    const expression = (...params: Parameters<Required<MapColumnArguments>['expression']>) =>
-      args
-        .expression?.(...params)
-        .pipe(take(1))
-        .toPromise() ?? Promise.resolve(null);
-    const columnId = args.id != null ? args.id : args.name;
+  fn(input, args) {
+    const id = args.id ?? args.name;
 
-    const columns = [...input.columns];
-    const rowPromises = input.rows.map((row) => {
-      return expression({
-        type: 'datatable',
-        columns,
-        rows: [row],
-      }).then((val) => ({
-        ...row,
-        [columnId]: val,
-      }));
-    });
+    return defer(() => {
+      const rows$ = input.rows.length
+        ? zip(
+            ...input.rows.map((row) =>
+              (
+                args.expression?.({
+                  type: 'datatable',
+                  columns: [...input.columns],
+                  rows: [row],
+                }) ?? of(null)
+              ).pipe(map((value) => ({ ...row, [id]: value })))
```

Comment:
i think we don't need the pipe here

Reply:
We do need that because the expression returns a value that we need to put into the row. The expression is an observable, so we have to use the `pipe` here to process the value upon subscription.